### PR TITLE
Device Binding / Verifier feature - SDKS-1747,  SDKS-2022 

### PR DESCRIPTION
**Pull request by jeyanthanperiyasamy**
# JIRA Ticket

Please link jira ticket here

# Description

Briefly describe the change and any information that would help speedup the review and testing.

# Definition of Done Checklist:

- [ ] Acceptance criteria is met.
- [ ] All tasks listed in the user story have been completed.
- [ ] Coded to standards.
- [ ] Ensure backward compatibility.
- [ ] API reference docs is updated.
- [ ] Unit tests are written.
- [ ] Integration tests are written.
- [ ] e2e tests are written.
- [ ] Functional spec is written/updated.
- [ ] contains example code snippets.
- [ ] Change log updated.
- [ ] Documentation story is created and tracked.
- [ ] Tech debts and remaining tasks are tracked in separated ticket(s).

### CODE REVIEW

File: forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/DeviceBindViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import android.content.Context
+import android.os.Parcelable
+import androidx.fragment.app.FragmentActivity
+import androidx.lifecycle.ViewModel
+import kotlinx.parcelize.Parcelize
+import org.forgerock.android.auth.InitProvider
+import org.forgerock.android.auth.callback.DeviceBindingAuthenticationType
+import org.json.JSONObject
+
+interface ViewModelHandler {
```
Comment (witrisna) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/DeviceBindViewModel.kt:12]: Do we need an interface? I think this interface name is too generic.

File: forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/DeviceBindViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            val key = users?.firstOrNull { it.userId == userId }
+            return key?.let { SingleKeyFound(it) } ?: NoKeysFound
+        }
+
+        return users?.let {
+            userKeys = it
+            when (it.size) {
+                0 -> NoKeysFound
+                1 -> SingleKeyFound(it.first())
+                else -> MultipleKeysFound(it, InitProvider.getCurrentActivityAsFragmentActivity())
```
Comment (witrisna) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/DeviceBindViewModel.kt:57]: Do we need the activity at this point?

File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+     *
+     * @param context  The Application Context
+     * @param listener The Listener to listen for the result
+     * @param authInterface Interface to find the Authentication Type
+     * @param encryptedPreference Persist the values in encrypted shared preference
+     */
+    protected open fun authenticate(userKey: UserKey,
+                             keyAware: KeyAware = KeyAware(userKey.userId),
+                             listener: FRListener<Void>) {
+        val authInterface =  getDeviceBindInterface(userKey, keyAware)
```
Comment (witrisna) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt:109]: Rename variable name to authenticator?

File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        viewModel.callback = {
+            authenticate(it, listener = listener)
+        }
+        fragment.show(activity.supportFragmentManager, DeviceBindFragment.TAG)
+    }
+
+    /**
+     * create the interface for the Authentication type(Biometric, Biometric_Fallback, none)
+     */
+    protected open fun getDeviceBindInterface(key: UserKey,  keyAware: KeyAware): Authenticator {
```
Comment (witrisna) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt:167]: Remane to getAuthenticator() ?

File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+                             listener: FRListener<Void>) {
+        val authInterface =  getDeviceBindInterface(userKey, keyAware)
+        if(authInterface.isSupported().not()) {
+            handleException(Unsupported(), listener)
+            return
+        }
+        try {
+            authInterface.authenticate(timeout ?: 60) { result ->
+                if (result is Success) {
+                    val privateKey = keyAware.getPrivateKey(userKey.keyAlias)
```
Comment (witrisna) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt:117]: Do we need the keyAware to retrieve the private key? If the authenticator get a hold of the keyAaware object, we may able to simplify the logic here by not passing the private key around.

File: forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/DeviceBindViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+            }
+        } ?: NoKeysFound
+    }
+}
+
+sealed class KeyFound
+data class SingleKeyFound(val key: UserKey): KeyFound()
+data class MultipleKeysFound(val keys: MutableList<UserKey>, val activity: FragmentActivity): KeyFound()
```
[the remark this answers is not in the document]
Comment (jeyanthanperiyasamy) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/DeviceBindViewModel.kt]: unlike enum Sealed classes are garbage collected , sealed class often used to share the state logic across viewmodel to Activity/Fragment , so i highly doubt its a leak, but i am going to move this activity to DeviceVerifier Callback

File: forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/DeviceBindViewModel.kt (hunk 1 above)
Comment (jeyanthanperiyasamy) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/DeviceBindViewModel.kt:12]: Are you taking about the purpose of this interface ?

File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt (hunk 5 above)
Comment (jeyanthanperiyasamy) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt:117]: Good point , we can wrap this in the authenticator

File: forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/DeviceBindViewModel.kt (hunk 1 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/DeviceBindViewModel.kt:12]: Yes, from the name ViewModelHandler, it sounds like handle all type of ViewModel, but this interface seems handling user's keys

File: forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/DeviceBindViewModel.kt (hunk 2 above)
Comment (jeyanthanperiyasamy) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/DeviceBindViewModel.kt:57]: actually i can move the logic to the execute method in the DeviceVerifierCallback,

File: forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/DeviceBindViewModel.kt (hunk 1 above)
Comment (jeyanthanperiyasamy) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/DeviceBindViewModel.kt:12]: i see your point, let me think. The ViewModel class is designed to have some business logic, store and manage UI-related data . In our case when we have multiple user keys we will display the Fragment  an we set the current selected key in the viewmodel  .  May be UserKeyViewModel , UserKeyViewModelStore . Do you have any other thoughts around this ?

File: forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/DeviceBindAuthenticators.kt [hunk 7]
```diff
@@ -62,10 +63,36 @@ interface Authenticator {
...
     /**
      * check biometric is supported
      */
     fun isSupported(): Boolean
+
+    /**
+     * Get the private key for the user
+     * @param keyAlias The user keyAlias
+     */
+    fun getPrivateKey(keyAlias: String): PrivateKey?
```
Comment (cerebrl) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/DeviceBindAuthenticators.kt:94]: This reads a bit scary. Is this inaccessible from outside this context because the function itself is defined on an `internal class`? Sorry if this is a silly question, but `getPrivateKey` just sounds worrying :)

File: forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/KeyAware.kt [hunk 8]
```diff
@@ -54,6 +53,12 @@ class KeyAware(private var userId: String) {
         return KeyPair(publicKey, privateKey, key)
     }
 
+
+    fun getPrivateKey(keyAlias: String = key): PrivateKey? {
```
Comment (cerebrl) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/KeyAware.kt:57]: What about this? Could this function be accessible to a bad actor?

**Review by cerebrl — COMMENTED** [review]
Looks good. I just left two comments about the `getPrivateKey` functions.

👍

File: forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/KeyAware.kt (hunk 8 above)
Comment (jeyanthanperiyasamy) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/KeyAware.kt:57]: @cerebrl actually its a good catch, the reason is they cannot directly access this , but they can inherit this class and invoke this method, so i am going to remove that method and directly consume the private key without exposing this

File: forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/DeviceBindAuthenticators.kt (hunk 7 above)
Comment (jeyanthanperiyasamy) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/DeviceBindAuthenticators.kt:94]: its a wonderful question. i am addressing this shortly

File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+    }
+
+    /**
+     * Display fragment to select a user key from the list
+     * @param activity activity to be used to display the Fragment
+     * @param userKeyService service to sort and fetch the keys stored in the device
+     * @param fragment to display the users list
+     * @param listener The Listener to listen for the result
+     */
+    protected open fun showKeysFragment(activity: FragmentActivity,
```
Comment (witrisna) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt:156]: As a developer, I would like to override this method to provide my UI, I am thinking of how we can achieve that:
```public class MyCallback extends DeviceSigningVerifierCallback {
    @Override
    protected void showKeysFragment(@NonNull FragmentActivity activity, @NonNull UserKeyService userKeyService, @NonNull DialogFragment fragment, @NonNull FRListener<Void> listener) {
    }
}
```

File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+                                        userKeyService: UserKeyService = UserDeviceKeyService(activity),
+                                        fragment: DialogFragment =  DeviceBindFragment(userKeyService),
+                                        listener: FRListener<Void>) {
+
+        userKeyService.callback = {
```
Comment (witrisna) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt:161]: Do we need the callback to associate with the UserKeyService?

**Review by witrisna — CHANGES_REQUESTED** [review]
Update copywrite header and javadoc

File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceBindingCallback.kt [hunk 11]
```diff
@@ -173,7 +173,7 @@ open class DeviceBindingCallback: AbstractCallback {
     /**
      * create the interface for the Authentication type(Biometric, Biometric_Fallback, none)
      */
-    open fun getDeviceBindInterface(): Authenticator {
+    protected open fun getAuthenticator(): Authenticator {
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceBindingCallback.kt:176]: This comment is not related to this PR, so no action item for now. But I think `Authenticator` is too generic for this interface. I would suggest rename it to something like `DeviceAuthenticator` or something more meaningful.

File: forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/DeviceBindFragment.kt [hunk 12]
```diff
@@ -0,0 +1,60 @@
+package org.forgerock.android.auth.devicebind
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/DeviceBindFragment.kt:1]: Missing Copyright

File: forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/KeyAware.kt [hunk 13]
```diff
@@ -11,7 +11,9 @@ import java.security.interfaces.RSAPublicKey
 /**
  * Helper class to generate and sign the keys
  */
-class KeyAware(private var userId: String) {
+internal class KeyAware(private var userId: String) {
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/KeyAware.kt:14]: Missing Copyright

File: forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/KeyAware.kt [hunk 14]
```diff
@@ -26,7 +28,6 @@ class KeyAware(private var userId: String) {
     private val key = getKeyAlias(userId)
 
     fun keyBuilder(): KeyGenParameterSpec.Builder {
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/KeyAware.kt:30]: doc?

File: forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/KeyAware.kt [hunk 15]
```diff
@@ -54,6 +55,12 @@ class KeyAware(private var userId: String) {
         return KeyPair(publicKey, privateKey, key)
     }
 
+
+    fun getSecureKey(keyAlias: String = key): PrivateKey? {
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/KeyAware.kt:59]: doc?

File: forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/UserDeviceKeyService.kt [hunk 16]
```diff
@@ -0,0 +1,70 @@
+package org.forgerock.android.auth.devicebind
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/UserDeviceKeyService.kt:1]: Missing Copyright

File: forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/UserDeviceKeyService.kt [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+import android.content.Context
+import android.os.Parcelable
+import androidx.fragment.app.FragmentActivity
+import androidx.lifecycle.ViewModel
+import kotlinx.parcelize.Parcelize
+import org.forgerock.android.auth.InitProvider
+import org.forgerock.android.auth.callback.DeviceBindingAuthenticationType
+import org.json.JSONObject
+
+interface UserKeyService {
```
[the remark this answers is not in the document]
Comment (rodrigoareis) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/UserDeviceKeyService.kt:12]: +1

File: forgerock-auth/src/test/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallbackTest.kt [hunk 18]
```diff
@@ -0,0 +1,212 @@
+package org.forgerock.android.auth.callback
```
Comment (rodrigoareis) [inline forgerock-auth/src/test/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallbackTest.kt:1]: Missing Copyright

File: forgerock-auth/src/test/java/org/forgerock/android/auth/devicebind/UserDeviceKeyServiceTest.kt [hunk 19]
```diff
@@ -0,0 +1,86 @@
+package org.forgerock.android.auth.devicebind
```
Comment (rodrigoareis) [inline forgerock-auth/src/test/java/org/forgerock/android/auth/devicebind/UserDeviceKeyServiceTest.kt:1]: Missing Copyright

File: samples/auth/src/main/res/values/strings.xml [hunk 20]
```diff
@@ -49,7 +49,7 @@
     <string name="forgerock_registration_service" translatable="false">Registration Tree2</string>
 
     <string name="asset_statements" translatable="false">           [{
-             "include": "https://openam.example.com/.well-known/assetlinks.json"
+             "include": "https://openam-forgerrock-sdks.forgeblocks.com/.well-known/assetlinks.json"
```
Comment (rodrigoareis) [inline samples/auth/src/main/res/values/strings.xml:52]: Should we revert it back to example URL?

**Review by rodrigoareis — CHANGES_REQUESTED** [review]
Overall changes looks good to me. Minor changes required

File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt (hunk 10 above)
Comment (jeyanthanperiyasamy) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt:161]: Fragment gets the userkeys from UserKeyService and set a selected key to UserKeyService and notify the Callback.  This is Similar to LiveData approach . we can also achieve this in multiple ways , we can also Directly fetch selected item from Fragment which may look cleaner

File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt (hunk 9 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt:156]: Would be good if we can allow developer to override with the following interface:
```java
    @Override
    protected void showKeysFragment(@NonNull FragmentActivity activity, @NonNull UserKeyService userKeyService ) {
        //somehow show my UI for all the userkeys and let the user pick one.
        UserKey userKey = userKeyService.getUserKeys().get(0);
        authenticate(userKey);
    }
```

File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt (hunk 10 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt:161]: Understand the idea, but reset the callback may be a concern, if the caller providers the callback for the userKeyService, the callback may be overwritten by this method.

Comment (jeyanthanperiyasamy) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt]: this is taken care

File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt (hunk 9 above)
Comment (jeyanthanperiyasamy) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt]: this is taken care

File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceBindingCallback.kt (hunk 11 above)
Comment (jeyanthanperiyasamy) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceBindingCallback.kt]: yeah it is more generic, DeviceAuthenticator should work

File: forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/UserDeviceKeyService.kt (hunk 17 above)
[the remark this answers is not in the document]
Comment (jeyanthanperiyasamy) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/UserDeviceKeyService.kt:13]: Done

File: forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/KeyAware.kt (hunk 13 above)
Comment (jeyanthanperiyasamy) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/KeyAware.kt:20]: thanks , since i updated the android studio all my copyright generators are gone , need to fix it back

**Review by witrisna — COMMENTED** [review]
Looks good to me, but please hold the merge for now.

**Review by witrisna — APPROVED** [review]
